windows-gaming-driver/src/setup/vfio.rs

```diff
@@ -79,6 +82,75 @@ pub fn setup(setup: &mut SetupConfig, machine: &mut MachineConfig) -> bool {
     true
 }
 
+fn select(machine: &mut MachineConfig) -> Result<()> {
+    let udev = Context::new().expect("Failed to create udev context");
+    let mut iter = Enumerator::new(&udev)?;
+    iter.match_subsystem("pci")?;
+    let pci_devs: Vec<_> = iter.scan_devices()?.map(PciDevice::new).collect();
+
+
+    // filter to the display controller class (0x03XXXX, udev drops the leading zero)
+    select_device(machine, "Which graphics card would you like to pass through?", &pci_devs,
+                  |x| x.pci_class.starts_with("3") && x.pci_class.len() == 5)?;
+    if ask::yesno("Would you like to pass aditional PCI devices?") {
+        println!();
+        println!("Please choose aditional pci devices. These will be passed through to the VM on boot.");
+        println!("Note: For this to work properly, the device has to support resetting.");
+        println!("The devices will be bound to the vfio-pci driver on VM startup and unbound when it quits.");
+        println!("You will not be able to use them on your host system while Windows is running.");
+        println!();
+
+        while select_device(machine, "Which device woud you like to pass through?", &pci_devs,
+                            |x| x.resettable)? { }
+
+        println!("Note that non-resettable passthrough is only known to work for GPUs.");
+        println!("With everything else you're completely on your own.");
+        println!("You will not be able to use these devices on your host system at all.");
+        println!();
+        if ask::yesno("Would you like to pass any non-resettable PCI devices?") {
+            // all non-resettable devices but no PCI bridges (because arch wiki says so)
+            while select_device(machine, "Which device do you want to pass through?", &pci_devs,
+                                |x| !x.resettable && x.pci_class != "60400")? { }
+        }
+    }
+    Ok(())
+}
+
+fn select_device<P: Fn(&PciDevice) -> bool>(machine: &mut MachineConfig,
+                                            question: &str,
+                                            devices: &[PciDevice],
+                                            predicate: P) -> Result<bool> {
+    let askable_devices: Vec<_> = devices.iter()
+        .filter(|x| predicate(x) && !machine.pci_devices.iter().any(|y| y.device == x.id)).collect();
+
+    println!("{}", devices.len());
+
+    for (i, dev) in askable_devices.iter().enumerate() {
+        println!("[{}]\t{}", i, dev);
+    }
+    println!("[{}]\tNone of the above.", askable_devices.len());
+
+    let selection = ask::numeric(question, 0..askable_devices.len()+1);
+
+    if let Some(selected_slot) = askable_devices.get(selection).map(|x| x.pci_slot.clone()) {
+        for device in devices.iter()
+            .filter(|x| x.pci_slot[..x.pci_slot.len()-1] == selected_slot[..selected_slot.len()-1]) {
```

Comment:
`selected_slot[..selected_slot.len()-1]` is the same as `selected_slot[..]`. Then it can also be all in the same line.